hydrus/data/resource_based_classes.py

```diff
@@ -0,0 +1,209 @@
+        inserted_object = database_class(**object_)
+    except TypeError as e:
+        # extract the wrong property name from TypeError object
+        wrong_property = e.args[0].split("'")[1]
+        raise PropertyNotFound(type_=wrong_property)
+    try:
+        session.add(inserted_object)
+        session.commit()
+    except Exception as e:
+        # catching any database contraint errors
+        session.rollback()
+        contraint_error = e.orig
+        raise DatabaseConstraintError(contraint_error)
+    return inserted_object.id
+
+
+def get_object(
+    query_info: Dict[str, str], session: scoped_session
+) -> Dict[str, str]:
+    """
+    Get the object from the database
+    :param query_info: Dict containing the id and @type of object that has to retrieved
+    :param session: sqlalchemy session
+    :return: dict of object with its properties
+    """
+    type_ = query_info["@type"]
+    id_ = query_info["id_"]
+    database_class = get_database_class(type_)
+    try:
+        object_ = (
+            session.query(database_class)
+            .filter(database_class.id == id_)
+            .one()
+        ).__dict__
+    except NoResultFound:
+        raise InstanceNotFound(type_=type_, id_=id_)
+    # Remove the unnecessary keys from the object retrieved from database
+    object_template = copy.deepcopy(object_)
+    object_template.pop("_sa_instance_state")
+    object_template.pop("id")
+    object_template["@type"] = query_info["@type"]
+    return object_template
+
+
+def delete_object(query_info: Dict[str, str], session: scoped_session) -> None:
+    """
+    Delete the object from the database
+    :param query_info: Dict containing the id and @type of object that has to retrieved
+    :param session: sqlalchemy session
+    """
+    type_ = query_info["@type"]
+    id_ = query_info["id_"]
+    database_class = get_database_class(type_)
+    try:
+        object_ = (
+            session.query(database_class)
+            .filter(database_class.id == id_)
+            .one()
+        )
+    except NoResultFound:
+        raise InstanceNotFound(type_=type_, id_=id_)
+    session.delete(object_)
+    try:
+        session.commit()
+    except Exception:
+        session.rollback()
+
+
+def update_object(
+    object_: Dict[str, Any],
+    query_info: Dict[str, str],
+    session: scoped_session,
+) -> str:
+    """
+    Update the object from the database
+    :param object_: Dict containing updated object properties
+    :param query_info: Dict containing the id and @type of object that has to retrieved
+    :param session: sqlalchemy session
+    :return: the ID of the updated object
+    """
+    # Keep the object as fail safe
+    old_object = get_object(query_info, session)
+    # Delete the old object
+    delete_object(query_info, session)
+    id_ = query_info["id_"]
+    # Try inserting new object
+    try:
+        object_["id"] = id_
+        d = insert_object(object_, session)
+    except Exception as e:
```

Comment:
Here generic exceptions are fine because we know that insert object will throw only certain exceptions.

Reply:
Understood :+1: